service/service.go

```diff
@@ -0,0 +1,181 @@
+	Producer     kafka.Producer
+	S3           *s3.S3
+}
+
+type inputFileAvailable struct {
+	FileURL    string `avro:"file_url"`
+	InstanceID string `avro:"instance_id"`
+}
+
+type dimensionExtracted struct {
+	FileURL    string `avro:"file_url"`
+	InstanceID string `avro:"instance_id"`
+}
+
+// ----------------------------------------------------------------------------
+
+// Start handles consumption of events ad produces new events
+func (svc *Service) Start() {
+	log.Info("application started", log.Data{"import-api-url": svc.ImportAPIURL})
+
+	c := make(chan os.Signal, 1)
+	signal.Notify(c, os.Interrupt, os.Kill)
+
+	healthChannel := make(chan bool)
+	exitChannel := make(chan bool)
+
+	go func() {
+		for {
+			select {
+			case <-c:
+				//Falls into this block when the service is shutdown to safely close the consumer
+				log.Info("safely closing consumer and producer on service shutdown", nil)
+				svc.Consumer.Closer <- true
+				svc.Producer.Closer <- true
+				exitChannel <- true
+				return
+			case message := <-svc.Consumer.Incoming:
+				m, err := readMessage(message.GetData())
+				if err != nil {
+					log.ErrorC("could not unmarshal message", err, log.Data{"instanceID": m.InstanceID})
+					continue
+				}
+				log.Info("received message", log.Data{"file-url": m.FileURL, "instance-id": m.InstanceID})
+
+				// Get csv from S3 bucket using m.S3URL
+				//s3Value := svc.s3Value
+				file, err := svc.S3.Get(m.FileURL)
+				if err != nil {
+					log.ErrorC("encountered error retrieving csv file", err, log.Data{"instanceID": m.InstanceID})
+					continue
+				}
+
+				log.Info("file successfully read from aws", log.Data{"instanceID": m.InstanceID})
+
+				csvReader := csv.NewReader(file)
+
+				// Scan and discard header row (for now) - the data rows contain sufficient information about the structure
+				if _, err := csvReader.Read(); err != nil {
+					log.ErrorC("encountered error immediately when processing header row", err, log.Data{"instanceID": m.InstanceID})
+				}
+
+				dimensionExtractedSchema := &avro.Schema{
+					Definition: schema.DimensionsExtracted,
+				}
+
+				producerMessage, err := dimensionExtractedSchema.Marshal(&dimensionExtracted{
+					FileURL:    m.FileURL,
+					InstanceID: m.InstanceID,
+				})
+
+				dimensions := make(map[string]string)
+				count := 0
+
+				// Iterate over csv file pulling out unique dimensions
+				for {
+					line, err := csvReader.Read()
+					if err == io.EOF {
+						break
+					}
+					if err != nil {
+						log.ErrorC("encountered error reading csv", err, log.Data{"instanceID": m.InstanceID, "csv_line": line})
+						break
+					}
+
+					dimensions, count, err = svc.sendRequests(m.InstanceID, dimensions, line, count)
+					if err != nil {
+						log.ErrorC("encountered error sending request to import API", err, log.Data{"instanceID": m.InstanceID, "csv_line": line})
+						break
+					}
+				}
+				log.Trace("count is:", log.Data{"instanceID": m.InstanceID, "count": count})
```

Comment:
It's not important, it was used to help debug whether dimensions map was being correctly used and matched results previously found for test file. It will be removed.